### PR TITLE
Fix <TracePathMatcher should match pattern "**" with paths end by "/">

### PR DESCRIPTION
**Pull request by DominikHubacek**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
@devkanro Please help on reviewing this new one.

**Comment by wu-sheng** [comment]
@DominikHubacek You need to fix CI, look like code style issue.

**Comment by wu-sheng** [comment]
@DominikHubacek It is better you could confirm all these edge cases. https://github.com/apache/skywalking-java/pull/72#issuecomment-975170969

**Comment by DominikHubacek** [comment]
> @DominikHubacek You need to fix CI, look like code style issue.

Will do

**Comment by DominikHubacek** [comment]
> @DominikHubacek It is better you could confirm all these edge cases. https://github.com/apache/skywalking-java/pull/72#issuecomment-975170969

I changed tests based on that comment, but only those which are checked(true), will add those unchecked (false) as well

**Comment by devkanro** [comment]
ok

**Comment by devkanro** [comment]
I have improved the state machine to support #72, check the comment to get more infomation.
@DominikHubacek @wu-sheng

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java [hunk 1]
```diff
@@ -84,6 +84,22 @@ private boolean normalMatch(String pat, int p, String str, int s) {
     private boolean wildcardMatch(String pat, int p, String str, int s) {
         char pc = safeCharAt(pat, p);
 
+        // End of pattern, check string only has zero or one '/' at end.
```
Comment (devkanro) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java:87]: If we get EOS(end of string) when we enter wildcard match state.
We just check the string only has zero or one '/' at end.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java [hunk 2]
```diff
@@ -122,9 +138,11 @@ private boolean wildcardMatch(String pat, int p, String str, int s) {
...
 
     private boolean multiWildcardMatch(String pat, int p, String str, int s) {
-        // End of pattern, just check the end of string is '/' quickly.
-        if (p >= pat.length() && s < str.length()) {
-            return str.charAt(str.length() - 1) != '/';
+        switch (safeCharAt(pat, p)) {
+            // End of pattern, just return true quickly.
+            case 0: return true;
+            // Skip next '/' for pattern to match zero path part.
+            case '/': p++;
```
Comment (devkanro) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java:145]: Add this branch to support `eureka/**/test` matches `eureka/test` (** means zero path part)

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java [hunk 3]
```diff
@@ -122,9 +138,11 @@ private boolean wildcardMatch(String pat, int p, String str, int s) {
     }
 
     private boolean multiWildcardMatch(String pat, int p, String str, int s) {
-        // End of pattern, just check the end of string is '/' quickly.
-        if (p >= pat.length() && s < str.length()) {
-            return str.charAt(str.length() - 1) != '/';
+        switch (safeCharAt(pat, p)) {
+            // End of pattern, just return true quickly.
+            case 0: return true;
```
Comment (devkanro) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java:143]: If we get EOS when we enter multi-wildcard match state, return true quickly, because `**` mean anything

**Comment by wu-sheng** [comment]
OK, so, this becomes a hybrid PR. @DominikHubacek Please recheck this PR, and if you don't feel comfortable about this PR will(if merged) be on your name, we could consider other options.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java (hunk 2 above)
Comment (DominikHubacek) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/FastPathMatcher.java:145]: > OK, so, this becomes a hybrid PR. @DominikHubacek Please recheck this PR, and if you don't feel comfortable about this PR will(if merged) be on your name, we could consider other options.

its ok, I checked if my changes still work as intended and fixed 1 test based on devkanro comment

**Comment by DominikHubacek** [comment]
> OK, so, this becomes a hybrid PR. @DominikHubacek Please recheck this PR, and if you don't feel comfortable about this PR will(if merged) be on your name, we could consider other options.

its ok, I checked if my changes still work as intended and fixed 1 test based on devkanro comment

**Comment by wu-sheng** [comment]
Could you two confirm this is good now?

**Review by wu-sheng — APPROVED** [review]
